- Description: fix channel path vary on for disconnected devices
  Symptom:     a ccw device does does not recover after disabled
               channel paths are reenabled on the service element
  Problem:     The service element does not notify the OS for
               enabling channel paths
  Solution:    Add more guesswork
  Problem-ID:  9971
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/cio/css.c	9 Jun 2004 11:21:53 -0000	1.72.2.3
+++ linux-2.5/drivers/s390/cio/css.c	14 Jul 2004 13:12:38 -0000	1.72.2.4
@@ -1,7 +1,7 @@
 /*
  *  drivers/s390/cio/css.c
  *  driver for channel subsystem
- *   $Revision: 1.72.2.3 $
+ *   $Revision: 1.72.2.4 $
  *
  *    Copyright (C) 2002 IBM Deutschland Entwicklung GmbH,
  *			 IBM Corporation
@@ -198,6 +198,20 @@
 	case CIO_GONE:
 		if (!sch) {
 			/* Never used this subchannel. Ignore. */
+			ret = 0;
+			break;
+		}
+		if (disc && (event == CIO_NO_PATH)) {
+			/*
+			 * Uargh, hack again. Because we don't get a machine
+			 * check on configure on, our path bookkeeping can
+			 * be out of date here (it's fine while we only do
+			 * logical varying or get chsc machine checks). We
+			 * need to force reprobing or we might miss devices
+			 * coming operational again. It won't do harm in real
+			 * no path situations.
+			 */
+			device_trigger_reprobe(sch);
 			ret = 0;
 			break;
 		}

